src/main/java/javaslang/collection/Stream.java

```diff
@@ -984,7 +991,11 @@ default int lastIndexOf(T element, int end) {
 
     @Override
     default Stream<T> removeLast(Predicate<T> predicate) {
-        return reverse().removeFirst(predicate).reverse();
+        if(findFirst(predicate).isDefined()) {
+            return reverse().removeFirst(predicate).reverse();
+        } else {
+            return this;
+        }
```

Comment:
I would do a `reverse().removeFirst(predicate)` instead of `findFirst` in order to not process elements 2 times, e.g.

```java
final Stream<T> removed = reverse().removeFirst(predicate);
return removed.length() == length() ? this : removed.reverse();
```